src/login/70-uaccess.rules.in

```diff
@@ -65,6 +65,7 @@ ENV{ID_REMOTE_CONTROL}=="1", TAG+="uaccess"
 
 # Gaming devices
 SUBSYSTEM=="input", ENV{ID_GAME_CONTROLLER}=="?*", TAG+="uaccess"
+SUBSYSTEM=="hidraw", ENV{ID_GAME_CONTROLLER}=="?*", TAG+="uaccess"
 SUBSYSTEM=="input", ENV{ID_INPUT_JOYSTICK}=="?*", TAG+="uaccess"
 
```

Comment:
should we really open up hidraw for all joysticks? shouldn't it just be the gaming devices that do more than regular joysticks?

i.e. opening hidraw should probably be restricted: if there's a proper input driver around handling these devices, we should push apps to use that, no? only if no full input driver exists we should open up hidraw, no?

Reply:
Please see the (4th) commit message. My understanding is that the problem is that the "proper" input driver is evdev, which involves some information loss: it's a semi-high-level interface that normalizes devices according to a Linux-specific abstraction layer, but that normalization removes information that is sometimes relevant. The difference is a bit like the difference between scancodes and keysyms, or the difference between raw and cooked terminals.

If the form of input you want is "the user is holding left on the d-pad while pressing the A button and pulling the right trigger 75% of the way down", then that's fine. However, if you're Wine passing through a HID device to a Windows game (which expects an interface that is basically equivalent to Linux hidraw devices), then the form of input you really want is more like "the user is pressing the 2nd and 7th buttons, while the 5th analog axis is at 75%", so that the order of the axes matches up with what's listed in the HID descriptor that was also passed through, and is identical to what the same device would do in Windows.

Wine is able to take an evdev device representing a gamepad and fake a semi-plausible HID device from it (basically by assuming it looks like an Xbox gamepad), but I don't think it's able to do that for more general game controllers like flight sticks, steering wheels and gyroscopes, because too much information has been lost in the translation from HID to evdev.

I think SDL and Steam Input have other reasons to prefer raw HID, but I'm not 100% clear on what those reasons are. @slouken might able to say more about this?

Reply:
> I think SDL and Steam Input have other reasons to prefer raw HID, but I'm not 100% clear on what those reasons are. @slouken might able to say more about this?

Steam and SDL use the raw HID device for PS4, PS5, Nintendo Switch, and Steam Controllers for direct report access to support advanced features in a consistent way across all platforms and kernel versions.

Reply:
> devices that do more than regular joysticks

That seems like it assumes that a "regular joystick" is something we can easily describe, which I don't think is really the case any more. Are you thinking of the gameport (MIDI-port) joysticks that you might have used 25 years ago, with 2 axes and some numbered buttons? Those are not actually very common or interesting any more.

The baseline level of controller complexity is an Xbox gamepad (6 axes and 15 buttons, if I'm counting correctly), and it goes up from there: in particular, recent Playstation gamepads show up in the evdev interface as three separate evdev devices (one for the functionality it shares with an Xbox gamepad, one for 6 axes of motion controls, and one for a laptop-style clickable trackpad), and it's up to user-space to figure out which ones belong to the same physical device and put them back together.

Reply:
> for direct report access to support advanced features in a consistent way across all platforms and kernel versions

What are those advanced features and why would a solution based on evdev not be consistent?

Reply:
> What are those advanced features

@slouken would likely have a better answer to this part than I do.

> why would a solution based on evdev not be consistent?

For SDL, this is mainly a trade-off about where to put the hardware abstraction layer and device-specific knowledge.

One of the goals of libraries like SDL is to present a cross-platform API to games and game-like programs. If it can implement that API by doing literally the same HID operations on Windows, macOS and Linux, then the behaviour of the hardware (including subtleties that games probably shouldn't rely on but almost certainly do, like the order in which axes and buttons are enumerated) is going to end up identical, and a bug-fix for one OS will immediately benefit the others too.

If SDL needs to do HID requests on Windows and macOS, but evdev for input / sysfs for LEDs / etc. on Linux (which ends up being translated into HID requests by the kernel, in a way that can differ between kernel releases), then the same version of SDL will have one set of behaviours on Windows and macOS, plus one or more different sets of behaviours on Linux (possibly kernel-dependent). Faced with that, game developers are very likely to test on Windows, see that their game works fine and call that good enough, even if it doesn't actually work as intended on commonly-used versions of Linux.

Having the hardware-specific knowledge in SDL (or, more generally, in user-space) means you can gain support for new game controllers by upgrading SDL, but not by upgrading the kernel; conversely, having the hardware-specific knowledge in the kernel means you can gain support for new game controllers by upgrading the kernel, but not by upgrading user-space. Mechanisms like the Steam Runtime and Flatpak make it quite straightforward to use an updated SDL, even on LTS distributions that continue to use an old kernel.

As I think I mentioned previously, one of several APIs for game controllers in Windows is basically HID ("dinput", emulated by Windows for Xbox controllers, or direct USB/Bluetooth HID for most others), so any game that wants to run on Windows needs some way to deal with HID *anyway*. In practice, basically all game developers want their game to be able to run on Windows, so we might as well take advantage of that.

The API that's available on Windows is also very relevant for Wine, which needs to be able to fake that API in terms of the information it can get from Linux. As I mentioned previously, when the Linux kernel converts lower-level HID into higher-level evdev, some information is lost (the application sees for example `BTN_SOUTH` and `BTN_EAST`, but doesn't know what order those appeared in for the lower-level HID interface), which makes it difficult for Wine to convert evdev back into HID in a way that will provide the same game-visible behaviour as Windows' "dinput" interface.

Reply:
chiming in here: evdev is imo woefully unsuitable for joysticks, full stop. as said above it's a lossy conversion from HID, it splits a single physical device into separate event nodes (that the evdev client then has to re-associate via guesswork) and it's linux only. The standardisation of events makes life easier for clients but that doesn't really work for joysticks which tend to push the limits of evdev (and adding new `EV_ABS` axes is virtually impossible without a major rework in the kernel and a new API).

 The *only* advantage left is that it has `EVIOCREVOKE`.

Reply:
> adding new EV_ABS axes is virtually impossible without a major rework in the kernel and a new API

That nicely explains why the PS3/PS4/PS5 controller is exposed as two or three separate evdev devices (6-axis joysticks/triggers, 6-axis accelerometer/gyro, and 2-axis trackpad on PS4/PS5) rather than piling everything into one device: the left stick, the trackpad and the accelerometer all want to be `ABS_X`/`ABS_Y`, the right stick and the gyro both want to be `ABS_RX`/`ABS_RY`, and the triggers, gyro and accelerometer all want `ABS_Z`/`ABS_RZ`.

> standardisation of events makes life easier for clients

Yes and no: some of the event maps are very explicit and obvious (for example I like the way the face buttons have been standardized as `BTN_NORTH`, etc. so that it's obvious that the upward-pointing face button should always be `BTN_NORTH` even though it's labelled Y on Xbox, triangle on Playstation and X on Nintendo), but the conventional mapping for the right stick on a gamepad is somewhat wrong (it repurposes `ABS_RX` and `ABS_RY`, which in theory are meant to measure rotation), and the conventional mapping for analogue triggers is just odd (a z-axis associated with the stick on the same side? really?)

Reply:
very much a case of hammers in desperate search for nails :) The inertia of getting changes in also means that some axes are abused because it's easier to just know that `ABS_FOO` is really "bar" then getting `ABS_BAR` added. This was true even before the multitouch support which mostly used up the remaining namespace - see also the wacom drivers which had some creative use of the abs space.

The separation into multiple event nodes is also a factor of the kernel hid drivers splitting per HID application collection - which mostly makes sense and absolves us from having to deal with e.g. a pen and multitouch on the same event node, without a good way of splitting the event streams correctly.

Joysticks, as usual, just tend to fall outside the easy-to-deal-with boundaries.